examples/exex/rollup/src/db.rs

```diff
@@ -0,0 +1,314 @@
+
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Get block by number.
+    pub fn get_block(&mut self, number: U256) -> eyre::Result<Option<SealedBlockWithSenders>> {
+        let block = self.connection().query_row::<String, _, _>(
+            "SELECT data FROM block WHERE number = ?",
+            (number.to_string(),),
+            |row| row.get(0),
+        );
+        match block {
+            Ok(data) => Ok(Some(serde_json::from_str(&data)?)),
+            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// Insert new account if it does not exist, update otherwise. The provided closure is called
+    /// with the current account, if it exists.
+    pub fn upsert_account(
+        &mut self,
+        address: Address,
+        f: impl FnOnce(Option<AccountInfo>) -> eyre::Result<AccountInfo>,
+    ) -> eyre::Result<()> {
+        let mut connection = self.connection();
+        let tx = connection.transaction()?;
+
+        let account = get_account(&tx, address)?;
+        let account = f(account)?;
+        tx.execute(
+            "INSERT INTO account (address, data) VALUES (?, ?) ON CONFLICT(address) DO UPDATE SET data = excluded.data",
+            (address.to_string(), serde_json::to_string(&account)?),
+        )?;
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Get account by address.
+    pub fn get_account(&mut self, address: Address) -> eyre::Result<Option<AccountInfo>> {
+        get_account(&self.connection(), address)
+    }
+}
+
+/// Get account by address using the database connection. Connection can be either
+/// [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_account<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<Option<AccountInfo>> {
+    match connection.deref().query_row::<String, _, _>(
+        "SELECT data FROM account WHERE address = ?",
+        (address.to_string(),),
+        |row| row.get(0),
+    ) {
+        Ok(account_info) => Ok(Some(serde_json::from_str(&account_info)?)),
+        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+        Err(e) => Err(e.into()),
+    }
+}
+
+/// Get all storages for the provided address using the database connection. Connection can be
+/// either [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_storages<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<Vec<(B256, U256)>> {
+    connection
+        .prepare("SELECT key, data FROM storage WHERE address = ?")?
+        .query((address.to_string(),))?
+        .mapped(|row| {
+            Ok((
+                B256::from_str(row.get_ref(0)?.as_str()?),
+                U256::from_str(row.get_ref(1)?.as_str()?),
+            ))
+        })
+        .map(|result| {
+            let (key, data) = result?;
+            Ok((key?, data?))
+        })
+        .collect()
+}
+
+impl reth::revm::Database for Database {
```

Comment:
would be cool for a service model that batches requests every 10-20ms to perhaps avoid doing single inserts all the time